### ISSUE TITLE
[BUG] OCIS_SHARING_PUBLIC_WRITEABLE_SHARE_MUST_HAVE_PASSWORD not correct

### ISSUE THREAD

**Issue by dj4oC** [owncloud/ios-app#1429]
Finding:
Read-Only Public Shares enforce passwords with oCIS

Expected behavior:

If `OCIS_SHARING_PUBLIC_SHARE_MUST_HAVE_PASSWORD == false` and `OCIS_SHARING_PUBLIC_WRITEABLE_SHARE_MUST_HAVE_PASSWORD == true` the password should only be enforced for writeable links, not for read only links.

Please fix.

Ref: https://owncloud.dev/services/frontend/#password-enforcement-for-all-public-links 

Thank you

ping: @DeepDiver1975 @hosy @felix-schwarz

**Comment by DeepDiver1975** [comment]
@kobergj are these env vars somehow exposed to the clients via e.g. capabilities or something? THX

**Comment by kobergj** [comment]
Yes of course. In capabilities: `"files_sharing.public.password.enforcedfor"`

**Comment by felix-schwarz** [comment]
@dj4oC This is [implemented by the iOS app](https://github.com/owncloud/ios-app/blob/0e17ba7247a7c03ac34ba0862064f305cf4b6536/ownCloudAppShared/Client/Sharing/ShareViewController.swift#L606), with the following logic:

- if `files_sharing.public.password.enforced` is `true`, it acts as a "catch-all" and a password is **always** required, irrespective of whether `files_sharing.public.password.enforced_for` exists or what it is set to.
- if `files_sharing.public.password.enforced` is `false`, a password is only required:

for | defined by permissions | if … is true
---|-------------------------|--------- 
read-write-delete links | `read` && `delete` && (`create` \|\| `update`) | `files_sharing.public.password.enforced_for.read_write_delete`
read-write links | `read` && (`create` \|\| `update`) | `files_sharing.public.password.enforced_for.read_write`
upload-only links | `create` | `files_sharing.public.password.enforced_for.upload_only`
read-only links | `read` | `files_sharing.public.password.enforced_for.read_only`

Can you share the full `files_sharing.public.password` part of the capabilities of an instance you see this issue with?

**Comment by dj4oC** [comment]
What I know is: if there is a read only share, I can remove the password in Web but I cannot on iOS. 

Web is working correctly.

**Comment by felix-schwarz** [comment]
@dj4oC I'll still need either the full `files_sharing.public.password` part of the capabilities of an instance you see this issue with - or (even better) a test account to investigate this further and work towards a solution.

**Comment by dj4oC** [comment]
> @dj4oC I'll still need either the full `files_sharing.public.password` part of the capabilities of an instance you see this issue with - or (even better) a test account to investigate this further and work towards a solution.

If you let me know how to get the information you need, I can provide them

**Comment by felix-schwarz** [comment]
Ok, after looking into it, I came up with these results:

- upon creating a link, the web UI also forces the user to set a password
- once the link is created, it is possible to remove the password via the context menu
- the server indicates in the capabilities that a password is enforced for `read_only`:
```json
"password": {
    "enforced_for": {
        "read_only": true,
        "read_write": true,
        "read_write_delete": true,
        "upload_only": true
    },
    "enforced": false
},
```

[To my understanding](https://github.com/owncloud/ios-app/issues/1429#issuecomment-2537253137), if `enforced` is `false`, a password is not **always** enforced, but only for the type of links where `enforced_for` is `true`. Is that assumption correct @kobergj?

If it is - and with `enforced_for.read_only` being `true`:
- the web client shouldn't offer the context menu option to remove the password 
- the server shouldn't allow removing the password
- the `OCIS_SHARING_PUBLIC_*` options need to be set differently - or interpreted differently

If it isn't:
- the web client shouldn't enforce setting a password on creation, either (for consistency)
- I need a reference on how else `enforced_for` and `enforced` should be interpreted by the app 🙂

(cc @LukasHirt)

**Comment by kobergj** [comment]
You are absolutely right. You can remove a password even if passwords are enforced. Imho this is a major bug. @dj4oC how can this be expected behaviour? I would suggest to fix this asap.

> I need a reference on how else enforced_for and enforced should be interpreted by the app

Here the answer is very simple: `enforced` will be always `false`. Therefore there is no need to use it in any way. `enforced_for` holds the only source of truth

**Comment by dj4oC** [comment]
No it was intended that you can **remove** the password for read only shares. So behaviour in web is correct.

**Comment by kobergj** [comment]
@dj4oC So to create a read-only share without passwords I only need to create one with password, then remove it. What's the point in enforcing a password then anyways? Bonus question: Why is this "security by default"?

@felix-schwarz sorry for spamming your ticket 😉

**Comment by dj4oC** [comment]
> @dj4oC So to create a read-only share without passwords I only need to create one with password, then remove it. What's the point in enforcing a password then anyways? Bonus question: Why is this "security by default"?

This is precisely the BSI/LSI obligation we have. It may be possible but the rule says: it is not allowed to CREATE a share without.

**Comment by dj4oC** [comment]
> Is that BSI/LSI obligation publicly accessible, so it could be referenced in documentation or code comments?

I've requested that document. But I am not full of hope, the classification will be removed

**Comment by felix-schwarz** [comment]
A solution to this issue is now implemented in https://github.com/owncloud/ios-app/pull/1443 and https://github.com/owncloud/ios-sdk/pull/138 respectively, by

1) adding support for a `block_password_removal` extension to the server's capabilities that would look as follows to achieve the requested behavior:

```
"password": {
    "enforced_for": {
        "read_only": true,
        "read_write": true,
        "read_write_delete": true,
        "upload_only": true
    },
    "block_password_removal": {
        "read_only": false,
        "read_write": false,
        "read_write_delete": false,
        "upload_only": false
    },
    "enforced": false
},
```

2) Assuming a default value of `false` for all `block_password_removal` settings to achieve the requested behavior when talking to a server not implementing or exposing the `block_password_removal` extension.

**Comment by DeepDiver1975** [comment]
take care of -> close

**Comment by DeepDiver1975** [comment]
@dj4oC ^ 👀  it is closed now 🔒
